modules/ml/src/test/java/org/apache/ignite/ml/selection/scoring/evaluator/EvaluatorTest.java

```diff
@@ -288,19 +289,24 @@ void evaluate(DecisionTreeClassificationTrainer trainer,
                 .withEncoderType(EncoderType.STRING_ENCODER)
                 .withEncodedFeature(1)
                 .withEncodedFeature(6) // <--- Changed index here
-                .fit(ignite,
+                .fit(
+                    testEnvBuilder(123L),
```

Comment:
Well, there are bunch of `with*` methods in extending classes and for API to be sane, i'd expect that `with` methods commute. If we'll make `PreprocessingTrainer withEnvironmentBuilder(...)` in `PreprocessingTrainer`, commutativity will not hold: `new BinarizationTrainer().withThreshhold(x).withEnvironmentBuilder(b)` will be correct, but `new BinarizationTrainer().withEnvBuilder(b).withThreshold(x)` will not compile because of return type of first `with`. Of course, we could return not just  `PreprocessingTrainer`, but in class signature  add `P extends PreprocessingTrainer`, and return this `P` in `withEnvironmentBuilder` and also make each subclass pass self into this param, but that'll make generics count `5` which is probably a little to much... But I personally not against it :)

Reply:
I think we should discuss such situation
I don't like both versions))) but we should choose between two evils)